gnsi/acltrie/trie.go

```diff
@@ -0,0 +1,256 @@
+		}
+		if node.elem.Name != path.Elem[depth].Name {
+			return false, nil
+		}
+		if depth == len(path.Elem)-1 && node.hasPolicy {
+			pathCmp := other
+			n := node
+			for i := depth; i >= 0; i-- {
+				elemCmp, err := comparePathElem(path.Elem[i], n.elem)
+				if err != nil {
+					return false, err
+				}
+				if pathCmp != other && elemCmp != pathCmp {
+					return false, fmt.Errorf("path is not consistently subset or superset of other rules")
+				}
+				if elemCmp != other {
+					pathCmp = elemCmp
+				}
+
+				n = n.parent
+			}
+		}
+		return true, nil
+	})
+	if err != nil {
+		return fmt.Errorf("policy path conflict: %v", err)
+	}
+
+	principal := r.GetUser()
+	policy := node.users
+	if _, isUser := r.GetPrincipal().(*pathzpb.AuthorizationRule_User); !isUser {
+		principal = r.GetGroup()
+		policy = node.groups
+	}
+
+	if err := policy.insert(principal, r.GetMode(), r.GetAction()); err != nil {
+		return fmt.Errorf("error inserting policy at %v: %v", r.Path, err)
+	}
+	node.hasPolicy = true
+
+	return nil
+}
+
+// walk explores the trie in breadth first order and invokes walkFn on every node.
+// To continue exploring children of the node, the walkFn must return true.
+func (t *Trie) walk(walkFn func(node *trieNode, depth int) (bool, error)) error {
+	type traversalNode struct {
+		node  *trieNode
+		depth int
+	}
+
+	queue := []*traversalNode{{node: t.root}}
+	for len(queue) > 0 {
+		front := queue[0]
+		queue = queue[1:]
+		for _, c := range front.node.children {
+			cont, err := walkFn(c, front.depth)
+			if err != nil {
+				return err
+			}
+			if cont {
+				queue = append(queue, &traversalNode{node: c, depth: front.depth + 1})
+			}
+		}
+	}
+	return nil
+}
+
+type compareResult int
+
+const (
+	other compareResult = iota
+	subset
+	superset
+)
+
+// comparePathElem compare two path elements a, b and returns:
+// subset: if every definite key in a is wildcard in b.
+// superset: if every wildcard key in b is non-wildcard in b.
+// other: all keys are the same or all keys are different.
+// error: not two keys are both subset and superset.
+// TODO: Make to more broadly useful compare that outputs subset, superset, equal, disjoint.
+func comparePathElem(a, b *gpb.PathElem) (compareResult, error) {
+	result := other
+	for k, aVal := range a.Key {
+		bVal, ok := b.Key[k]
+		switch {
+		case aVal == bVal:
+			continue
+		case aVal == "*" && !ok:
+			continue
+		case aVal == "*":
+			if result == subset {
+				return other, fmt.Errorf("path %v is not consistently a superset of %v", a, b)
+			}
+			result = superset
+		case bVal == "*" || !ok:
+			if result == superset {
+				return other, fmt.Errorf("path %v is not consistently a subset of %v", a, b)
+			}
+			result = subset
+		}
+	}
+	for k, bVal := range b.Key {
+		_, ok := a.Key[k]
+		if !ok && bVal != "*" { // If a contains an implicit wildcard.
+			if result == subset {
+				return other, fmt.Errorf("path %v is not consistently a superset of %v", a, b)
+			}
+		}
```

Comment:
missing `result = superset` at end of outer if statement? A missing test case?

Reply:
done and added case